src/App/src/Handler/HomePageHandler.php

```diff
@@ -28,7 +27,7 @@ public function __construct(Router\RouterInterface $router, Template\TemplateRen
         $this->template = $template;
     }
 
-    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) : ResponseInterface
+    public function handle(ServerRequestInterface $request) : ResponseInterface
```

Comment:
Although I like the idea and in most circumstances the RequestHandlerInterface is sufficient, I have one issue with using the RequestHandlerInterface instead if the MiddlewareInterface.  For example, the zend-session-authentication needs the MiddlewareInterface as it needs the delegate:

```php
        if ($request->getMethod() === 'POST') {
            $this->loginForm->setData($request->getParsedBody());
            if ($this->loginForm->isValid()) {
                // $handler is next in the pipeline = Zend\Expressive\Authentication\AuthenticationMiddleware
                $response = $handler->handle($request);
                if ($response->getStatusCode() !== 301) {
                    return new RedirectResponse('/');
                }
 
                $error = 'Login Failure, please try again';
            }
        }
```

I've got it from @samsonasik blog post, so I'm not sure if it should be used differently. But I guess it's being used in more situations where the action / handler isn't the last one in the pipeline.

Reply:
In these particular cases, we're not delegating; they are terminal (always return a response). As such, using a handler for each of them is appropriate.

In the case of components such as zend-session-authentication, those are properly creating middleware, as they will delegate as necessary. The `$handler` in the above sample is the one provided to the middleware, and would not be, for instance, this particular handler (though this handler may eventually be what returns the response!).

The point is: use `MiddlewareInterface` if what you're creating will call on the `$handler` at any point; use `RequestHandlerInterface` if it **never** will.

Reply:
I would think to use MiddlewareInterface everywhere for consistency.

> The point is: use MiddlewareInterface if what you're creating will call on the $handler at any point; use RequestHandlerInterface if it never will.

This makes a lot of sense though. Thanx for the explanation. We might want to add this in the docs in a best practice section or so.